sqlmesh/core/plan/builder.py

```diff
@@ -673,6 +719,95 @@ def _ensure_new_env_with_changes(self) -> None:
                 "No changes were detected. Make a change or run with --include-unmodified to create a new environment without changes."
             )
 
+    def _upstream_directly_modified(
+        self, s_id: SnapshotId, directly_modified: t.Set[SnapshotId], dag: DAG[SnapshotId]
+    ) -> t.List[SnapshotId]:
+        return [
+            upstream_id for upstream_id in dag.upstream(s_id) if upstream_id in directly_modified
+        ]
+
+    def _validate_destructive_change(
+        self,
+        s_id: SnapshotId,
+        upstream_snapshot_ids: t.List[SnapshotId] = [],
+        force_check: bool = False,
+    ) -> None:
+        """Determine if a schema change requires a destructive operation"""
+        s_id_snapshot = t.cast(Snapshot, self._context_diff.snapshots.get(s_id))
+
+        if (
+            s_id_snapshot.is_model
+            and s_id_snapshot.name not in self._allow_destructive_models
+            and not s_id_snapshot.model.on_destructive_change.is_ignore
+        ):
+            warning_msg = f"PLAN TIME CHECK: Plan results in a destructive change to forward-only model '{s_id_snapshot.name}'s schema."
+            error_msg = f"{warning_msg} To allow this, change the model's `on_destructive_change` setting to `warn` or `ignore` or include it in the plan's `--allow-destructive-model` option."
+
+            def _raise_or_warn() -> None:
+                if s_id_snapshot.model.on_destructive_change.is_error:
+                    raise PlanError(error_msg)
+                logger.warning(warning_msg)
+
+            subdag = set([s_id, *upstream_snapshot_ids])
+            subdag_to_process = subdag.copy()
```

Comment:
i wonder if this method could be more succint something like

subdag = set()

for id in (s_id, *upstream):
  if ...:
    subdag.add(id)